### DISCUSSION TITLE
0.19.0/detection/tools/line_zone/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1125] [Docs]
# 0.19.0/detection/tools/line_zone/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/0.19.0/detection/tools/line_zone/

**Comment by TaHa8Farag** [comment]
How can I change the place where the number is written from above or below the line to anywhere in the video?

**Comment by casperdias** [comment]
Hey, I have a question.
Can you provide an example using that line_zone
IF people_crossline
    Do Something
??
Thx

**Reply by SkalskiP** [reply]
Take a look here. That's probably the simplest case. `someone crossed` will get printed only if someone crossed the line.

```python
import supervision as sv
from ultralytics import YOLO

model = YOLO(<SOURCE_MODEL_PATH>)
tracker = sv.ByteTrack()
frames_generator = sv.get_video_frames_generator(<SOURCE_VIDEO_PATH>)
start, end = sv.Point(x=0, y=1080), sv.Point(x=3840, y=1080)
line_zone = sv.LineZone(start=start, end=end)

for frame in frames_generator:
    result = model(frame)[0]
    detections = sv.Detections.from_ultralytics(result)
    detections = tracker.update_with_detections(detections)
    crossed_in, crossed_out = line_zone.trigger(detections)
    if np.any(crossed_in) or np.any(crossed_out):
        print("someone crossed")
```

**Reply by casperdias** [reply]
TY, i tried it and it work. I have another question.
I try to crop and save image of someone who crossed, but when i try it it save all detection image not just who crossed??

**Reply by SkalskiP** [reply]
Sure, we can help you with that as well :)

```python
import supervision as sv
from ultralytics import YOLO

model = YOLO(<SOURCE_MODEL_PATH>)
tracker = sv.ByteTrack()
frames_generator = sv.get_video_frames_generator(<SOURCE_VIDEO_PATH>)
start, end = sv.Point(x=0, y=1080), sv.Point(x=3840, y=1080)
line_zone = sv.LineZone(start=start, end=end)

for frame in frames_generator:
    result = model(frame)[0]
    detections = sv.Detections.from_ultralytics(result)
    detections = tracker.update_with_detections(detections)
    crossed_in, crossed_out = line_zone.trigger(detections)
    if np.any(crossed_in) or np.any(crossed_out):
        detections_crossed_in = detections[crossed_in]
        for xyxy in detections_crossed_in.xyxy:
            crop = sv.crop_image(frame, xyxy)
```

**Comment by SudhanshuBlaze** [comment]
I am working with [this traffic video](https://artlist.io/stock-footage/clip/kazakhstan-traffic-cars-road/741755), and I want to classify and count different classes of vehicles crossing a line counter. I would also like to have a preview, where you can see the video processed frame by frame and not have to wait until the video is completely processed.

Current Code:

```
import supervision as sv
from ultralytics import YOLO

model = YOLO("yolov9e.pt")
tracker = sv.ByteTrack()
frames_generator = sv.get_video_frames_generator("resources/741755_Kazakhstan Traffic Cars Road_By_Danil_Nevsky_Artlist_HD.mp4")
start, end = sv.Point(x=0, y=1080), sv.Point(x=3840, y=1080)
line_zone = sv.LineZone(start=start, end=end)

# Initialize counters
crossed_count_in = 0
crossed_count_out = 0


for frame in frames_generator:
    result = model(frame)[0]
    detections = sv.Detections.from_ultralytics(result)
    detections = tracker.update_with_detections(detections)
    
    crossed_in, crossed_out = line_zone.trigger(detections)
    if np.any(crossed_in):
        crossed_count_in += np.sum(crossed_in)
    if np.any(crossed_out):
        crossed_count_out += np.sum(crossed_out)

    annotated_frame = bounding_box_annotator.annotate(frame.copy(), detections)

    # Annotate counts on the frame
    cv2.putText(annotated_frame, f"Crossed In: {crossed_count_in}", (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
    cv2.putText(annotated_frame, f"Crossed Out: {crossed_count_out}", (50, 100), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)

```


Goal which I want to achieve (with supervision):
<img width="1526" alt="Screenshot 2024-05-30 at 15 41 14" src="https://github.com/ultralytics/ultralytics/assets/63813872/b2e6bf3d-6cfd-43c7-b417-a40d0bc9d338">

The results in this image has been achieved with the help of `ultralytics` library, but it does not have a `Detection Smoother`, due to which I want to achieve the similar results with `Supervision library.

**Comment by Kutayykilic** [comment]
I don't think it's super clear how trigger anchors work. Can you give an example on how to change it to two points? Or possibly the center point of the square?

**Reply by LinasKo** [reply]
Hi @Kutayykilic 👋 

The basic idea is simple: since every detection can be of different size, we provide methods to automatically compute the location for points of reference (anchors) when it comes to crossing into zones.

In our implementation, for an object to cross a lind / PolygonZone, ALL of its anchors must cross it.

You can find the values of anchors here in [sv.Position](https://github.com/roboflow/supervision/blob/3d5437b96090133b1325a5c385cc53f283d80083/supervision/geometry/core.py#L9)

When creating LineZone, it accepts a `triggering_anchors` argument, which is simply a list of `sv.Position`.

Therefore, if you pass in `triggering_anchors=[Position.CENTER]`, your LineZone will work with the centerpoint of the detected boxes.
